events/nats_publish.go

```diff
@@ -0,0 +1,204 @@
+
+	reqMsg, err := newNatsMessage(c, topic, message)
+	if err != nil {
+		return nil, err
+	}
+
+	c.logger.Debugf("publishing auth relation request message to topic %s", topic)
+
+	respMsg, err := reqMsg.request(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return respMsg, nil
+}
+
+// PublishChange publishes a ChangeMessage.
+func (c *NatsConnection) PublishChange(ctx context.Context, topic string, message ChangeMessage) (Message[ChangeMessage], error) {
+	ctx, span := newEventTraceSpan(ctx, c.tracer, "events.nats.PublishAuthRelationshipRequest", topic, message)
+
+	defer span.End()
+
+	if err := message.Validate(); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return nil, err
+	}
+
+	// Propagate trace context into the message for the subscriber
+	var mapCarrier propagation.MapCarrier = make(map[string]string)
+
+	otel.GetTextMapPropagator().Inject(ctx, mapCarrier)
+
+	message.TraceContext = mapCarrier
+
+	topic = c.buildPublishSubject("changes", message.EventType, topic)
+
+	message.Source = c.cfg.Source
+
+	if message.ActorID == gidx.NullPrefixedID {
+		id, ok := ctx.Value(echojwtx.ActorCtxKey).(string)
+		if ok {
+			message.ActorID = gidx.PrefixedID(id)
+		} else {
+			message.ActorID = "unknown-actor"
+		}
+	}
+
+	span.SetAttributes(
+		attribute.String(
+			"events.actor_id",
+			message.ActorID.String(),
+		),
+	)
+
+	msg, err := newNatsMessage(c, topic, message)
+	if err != nil {
+		return nil, err
+	}
+
+	c.logger.Debugf("publishing change message to topic %s", topic)
+
+	if err = msg.publish(); err != nil {
+		return msg, err
+	}
+
+	return msg, nil
+}
+
+// PublishEvent publishes an EventMessage.
+func (c *NatsConnection) PublishEvent(ctx context.Context, topic string, message EventMessage) (Message[EventMessage], error) {
+	ctx, span := newEventTraceSpan(ctx, c.tracer, "events.nats.PublishAuthRelationshipRequest", topic, message)
+
+	defer span.End()
+
+	if err := message.Validate(); err != nil {
+		span.RecordError(err)
+		span.SetStatus(codes.Error, err.Error())
+
+		return nil, err
+	}
+
+	// Propagate trace context into the message for the subscriber
+	var mapCarrier propagation.MapCarrier = make(map[string]string)
+
+	otel.GetTextMapPropagator().Inject(ctx, mapCarrier)
+
+	message.TraceContext = mapCarrier
+
+	topic = c.buildPublishSubject("events", message.EventType, topic)
+
+	msg, err := newNatsMessage(c, topic, message)
+	if err != nil {
+		return nil, err
+	}
+
+	c.logger.Debugf("publishing event message to topic %s", topic)
+
+	if err = msg.publish(); err != nil {
+		return msg, err
+	}
+
+	return msg, nil
+}
+
+func newEventTraceSpan(ctx context.Context, tracer trace.Tracer, spanName, subjectType string, message any) (context.Context, trace.Span) {
+	var (
+		subjectID   string
+		eventType   string
+		eventSource string
+	)
+
+	switch msg := message.(type) {
+	case AuthRelationshipRequest:
+		subjectID = msg.SubjectID.String()
+		eventType = "request"
```

Comment:
Do we want to hard code this to "request" ? Wouldn't we lose the context of whether this was a delete vs a write?

Reply:
🤔 I'm not sure why i did that. I definitely remember thinking about that. Thanks!